content/blog/gsoc-25-netjsongraph.rst

```diff
@@ -0,0 +1,225 @@
+         // Offset in pixel space
+         const offsetPoint = [
+           basePoint.x + separationPx * Math.cos(angle),
+           basePoint.y + separationPx * Math.sin(angle),
+         ];
+         // Convert back to lat/lng for display
+         const offsetLatLng = self.leaflet.containerPointToLatLng(offsetPoint);
+         centroidLng = offsetLatLng.lng;
+         centroidLat = offsetLatLng.lat;
+       }
+
+4.  **Final Repulsion Pass:** To handle edge cases, a final, simple force-directed repulsion step treats all clusters and nearby single nodes as physical circles and gently pushes them apart if they overlap.
+
+Feature: Displaying Connected Clients on Nodes
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+- **The Problem:** The library lacked a built-in, flexible way to visualize secondary data points on nodes, such as the number of connected WiFi clients.
+
+(Placeholder for GIF showing nodes with and without the client overlay)
+
+- **The Mechanism & Logic:** I developed this feature as a modular overlay in ``src/js/netjsongraph.clients.js``.
+
+- **Public API:** The feature is exposed via an easy-to-use ``attachClientsOverlay`` method.
+
+    .. code-block:: javascript
+
+       // file: public/example_templates/netjsongraph-wifi-clients.html
+       const graph = new NetJSONGraph(data, {
+         onReady() {
+           // Attach reusable overlay plugin
+           this.attachClientsOverlay({ radius: 5, gap: 3 });
+         }
+       });
+
+- **Orbital Placement Algorithm:** To display clients, I implemented a spiral algorithm that places dots in expanding concentric rings. The number of dots per ring is calculated dynamically based on its circumference.
+
+    .. code-block:: javascript
+
+       // file: src/js/netjsongraph.clients.js (inside placeOrbit)
+       for (let orbit = 0; i < total; orbit += 1) {
+         const distance = Math.max(0.1, startDistance + orbit * 2 * radius * a);
+         // Calculate how many dots fit on the current ring's circumference
+         const n = Math.max(1, Math.floor((Math.PI * distance) / (a * radius)));
+         const delta = total - i;
+
+         for (let j = 0; j < Math.min(delta, n); j += 1, i += 1) {
+           const angle = ((2 * Math.PI) / n) * j;
+           const x = centerX + distance * Math.cos(angle);
+           const y = centerY + distance * Math.sin(angle);
+           overlay.add(new g.Circle(/* ... */));
+         }
+       }
+
+- **GUI Integration:** The sidebar was enhanced to recursively render nested data, allowing it to display detailed information for each client (e.g., MAC address).
+
+Feature: Zoom-Dependent Label Visibility
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+- **The Problem:** In both map and graph views, displaying all node labels simultaneously at low zoom levels results in a cluttered and unreadable visualization.
+
+(Placeholder for GIF showing labels appearing/disappearing on zoom)
```

Comment:
Here I think it would be great to have before and after GIFs. Showing what happened before when zooming out (labels overlapping), while in the new versions labels are hidden.